### PR TITLE
Use clinical data to draw scatter plot

### PR DESCRIPTION
**Pull request by zhx828**
# What? Why?
Fix new study view in issue https://github.com/cBioPortal/cbioportal/issues/4650 

# Checks
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!
- [ ] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Airbnb React/JSX Style guide](https://github.com/airbnb/javascript/tree/master/react).
- [ ] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [GifGrabber](http://www.gifgrabber.com/).

# Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). If you are part of the
cBioPortal organization, notify the approprate team (remove inappropriate):

@cBioPortal/frontend

If you are not part of the cBioPortal organization look at who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them here:

### CODE REVIEW

File: src/pages/studyView/StudyViewPageStore.ts [hunk 1]
```diff
@@ -672,76 +672,46 @@ export class StudyViewPageStore {
...
-                if (mutationCount && fga) {
-                    data.push({
-                        x: fga.value,
-                        y: mutationCount.mutationCount,
-                        studyId: mutationCount.studyId,
-                        sampleId: mutationCount.sampleId,
-                        patientId: mutationCount.patientId,
-                        uniqueSampleKey: mutationCount.uniqueSampleKey
+            let sampleScatterData: { [id: string]: { [id: string]: ClinicalData } } = {};
+            _.each(data, datum => {
```
Comment (kalletlak) [inline src/pages/studyView/StudyViewPageStore.ts]: may be `_.groupBy`

Comment (zhx828) [reply src/pages/studyView/StudyViewPageStore.ts]: groupBy does not work in this case, and I do want to use one for loop for constructing the data.

File: src/pages/studyView/StudyViewPageStore.ts [hunk 2]
```diff
@@ -377,12 +372,17 @@ export class StudyViewPageStore {
...
         }
 
-        _chartMetaSet[UniqueKey.MUTATION_COUNT_CNA_FRACTION] = {
-            uniqueKey: UniqueKey.MUTATION_COUNT_CNA_FRACTION,
-            chartType: ChartType.SCATTER,
-            displayName: 'Mutation count Vs. CNA',
-            description: ''
-        };
+        const allAttrs = this.clinicalAttributes.result.map(attr => attr.clinicalAttributeId);
+        if (_.every(['MUTATION_COUNT', 'FRACTION_GENOME_ALTERED'], attr => {
```
[the remark this answers is not in the document]
Comment (zhx828) [reply src/pages/studyView/StudyViewPageStore.ts]: I think so. this.clinicalAttributes is not related to study.

File: src/pages/studyView/StudyViewPageStore.ts [hunk 3]
```diff
@@ -377,12 +372,17 @@ export class StudyViewPageStore {
             };
         }
 
-        _chartMetaSet[UniqueKey.MUTATION_COUNT_CNA_FRACTION] = {
-            uniqueKey: UniqueKey.MUTATION_COUNT_CNA_FRACTION,
-            chartType: ChartType.SCATTER,
-            displayName: 'Mutation count Vs. CNA',
-            description: ''
-        };
+        const allAttrs = this.clinicalAttributes.result.map(attr => attr.clinicalAttributeId);
```
[the remark this answers is not in the document]
Comment (zhx828) [reply src/pages/studyView/StudyViewPageStore.ts]: Hm you mean combine with the following if condition? Yeah, good idea, I could do that.

File: src/pages/studyView/StudyViewPageStore.ts [hunk 4]
```diff
@@ -672,76 +672,46 @@ export class StudyViewPageStore {
...
+                        studyId: datum.MUTATION_COUNT.studyId,
+                        sampleId: datum.MUTATION_COUNT.sampleId,
+                        patientId: datum.MUTATION_COUNT.patientId,
+                        uniqueSampleKey: datum.MUTATION_COUNT.uniqueSampleKey
                     });
                 }
-            }
-            return Promise.resolve(data);
+            });
+            return Promise.resolve(result);
```
Comment (kalletlak) [inline src/pages/studyView/StudyViewPageStore.ts]: you could directly send `result`

File: src/pages/studyView/StudyViewPageStore.ts [hunk 5]
```diff
@@ -672,76 +672,46 @@ export class StudyViewPageStore {
...
-                const fga = sampleToFga[sampleKey];
-                if (mutationCount && fga) {
-                    data.push({
-                        x: fga.value,
-                        y: mutationCount.mutationCount,
-                        studyId: mutationCount.studyId,
-                        sampleId: mutationCount.sampleId,
-                        patientId: mutationCount.patientId,
-                        uniqueSampleKey: mutationCount.uniqueSampleKey
+            let sampleScatterData: { [id: string]: { [id: string]: ClinicalData } } = {};
```
Comment (adamabeshouse) [inline src/pages/studyView/StudyViewPageStore.ts]: can you name the two `id:string` more specifically so its clearer what the data structure is?

File: src/pages/studyView/StudyViewPageStore.ts [hunk 6]
```diff
@@ -672,76 +672,46 @@ export class StudyViewPageStore {
...
+            _.each(data, datum => {
+                if (!_.has(sampleScatterData, datum.uniqueSampleKey)) {
+                    sampleScatterData[datum.uniqueSampleKey] = {};
+                }
+                sampleScatterData[datum.uniqueSampleKey][datum.clinicalAttributeId] = datum;
+            })
+
+            let result: any = [];
+            _.each(sampleScatterData, datum => {
+                if (_.isObject(datum['MUTATION_COUNT']) && _.isObject(datum['FRACTION_GENOME_ALTERED'])) {
```
Comment (adamabeshouse) [inline src/pages/studyView/StudyViewPageStore.ts]: I think its safer to declare some const MUTATION_COUNT="MUTATION_COUNT" and use that, same with FGA, so we dont have any hidden typo errors

File: src/pages/studyView/StudyViewPageStore.ts (hunk 4 above)
Comment (adamabeshouse) [reply src/pages/studyView/StudyViewPageStore.ts]: yes since its async function ^^

**Comment by zhx828** [comment]
@kalletlak  @adamabeshouse updated, could you take a look? Thanks.

File: src/pages/studyView/StudyViewPageStore.ts [hunk 7]
```diff
@@ -377,12 +375,24 @@ export class StudyViewPageStore {
...
-        };
+        const scatterRequiredParams = _.reduce(this.clinicalAttributes.result, (acc, next) => {
+            if (MUTATION_COUNT === next.clinicalAttributeId) {
+                acc[MUTATION_COUNT] = true
+            }
+            if (FRACTION_GENOME_ALTERED === next.clinicalAttributeId) {
+                acc[FRACTION_GENOME_ALTERED] = true
+            }
+            return acc;
+        }, {MUTATION_COUNT: false, FRACTION_GENOME_ALTERED: false});
```
Comment (adamabeshouse) [inline src/pages/studyView/StudyViewPageStore.ts]: you need to do `[MUTATION_COUNT]:false, [FRACTION_GENOME_ALTERED]:false` to use the constants

File: src/pages/studyView/StudyViewPageStore.ts [hunk 8]
```diff
@@ -672,76 +682,43 @@ export class StudyViewPageStore {
...
-                        x: fga.value,
-                        y: mutationCount.mutationCount,
-                        studyId: mutationCount.studyId,
-                        sampleId: mutationCount.sampleId,
-                        patientId: mutationCount.patientId,
-                        uniqueSampleKey: mutationCount.uniqueSampleKey
-                    });
+            return _.reduce(_.groupBy(data, datum => datum.uniqueSampleKey), (acc, data) => {
+                if (data.length == 2) { // 2 => number of attribute ids
+                    let _datum: { [id: string]: string | number } = {
```
Comment (adamabeshouse) [inline src/pages/studyView/StudyViewPageStore.ts]: its not a map so I think the type should be more specific than index type, for readability. ie lets declare a type up top `type MutationCountVsFGADatum = { studyId:string, ....}`

File: src/pages/studyView/StudyViewPageStore.ts (hunk 7 above)
Comment (zhx828) [reply src/pages/studyView/StudyViewPageStore.ts]: Ah good catch.

File: src/pages/studyView/StudyViewPageStore.ts [hunk 9]
```diff
@@ -672,76 +691,45 @@ export class StudyViewPageStore {
...
+                        } else if (datum.clinicalAttributeId === FRACTION_GENOME_ALTERED) {
+                            _datum.x = Number(datum.value)
+                        }
                     });
+                    acc.push(_datum)
                 }
-            }
-            return Promise.resolve(data);
+                return acc
+            }, [] as { [id: string]: string | number }[]);
```
Comment (adamabeshouse) [inline src/pages/studyView/StudyViewPageStore.ts]: `[] as MutationCountVsFGADatum[]`

Comment (zhx828) [reply src/pages/studyView/StudyViewPageStore.ts]: Done

**Comment by zhx828** [comment]
@inodb all the checks are failing for this PR due to the backend change I think. Because the testing instance with rc is not updated?

**Comment by ersinciftci** [comment]
@zhx828 no, it's because you update the api client in this PR, but not all pages are fixed, so I'm doing this now in your branch, will push here soon.

**Pull request merged**